modules/images/webp-uploads/load.php

```diff
@@ -5,40 +5,294 @@
- * @return string The new output format mapping.
+ * @param array $metadata      An array with the metadata from this attachment.
+ * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
+ *
+ * @return array An array with the updated structure for the metadata before is stored in the database.
  */
-function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
-	// Only enable if the server supports WebP.
-	if ( ! wp_image_editor_supports( array( 'mime_type' => 'image/webp' ) ) ) {
-		return $output_format;
+function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
+	// This should take place only on the JPEG image.
+	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
+
+	// Not a supported mime type to create the sources property.
+	if ( ! array_key_exists( get_post_mime_type( $attachment_id ), $valid_mime_transforms ) ) {
+		return $metadata;
+	}
+
+	$file = get_attached_file( $attachment_id, true );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return $metadata;
+	}
+
+	$dirname     = pathinfo( $file, PATHINFO_DIRNAME );
+	$image_sizes = array();
+	if ( array_key_exists( 'sizes', $metadata ) && is_array( $metadata['sizes'] ) ) {
+		$image_sizes = $metadata['sizes'];
+	}
+
+	$sizes = array();
+
+	foreach ( wp_get_registered_image_subsizes() as $size_name => $properties ) {
+		// This image size does not exist on the defined sizes.
+		if ( ! array_key_exists( $size_name, $image_sizes ) || ! is_array( $image_sizes[ $size_name ] ) ) {
+			continue;
+		}
+
+		$current_size = $image_sizes[ $size_name ];
+		$sources      = array();
+		if ( array_key_exists( 'sources', $current_size ) && is_array( $current_size['sources'] ) ) {
+			$sources = $current_size['sources'];
+		}
+
+		// Try to find the mime type of the image size.
+		$current_mime = '';
+		if ( array_key_exists( 'mime-type', $current_size ) ) {
+			$current_mime = $current_size['mime-type'];
+		} elseif ( array_key_exists( 'file', $current_size ) ) {
+			$current_mime = wp_check_filetype( $current_size['file'] )['type'];
+		}
+
+		if ( empty( $current_mime ) ) {
+			$sizes[ $size_name ] = $current_size;
+			continue;
+		}
+
+		$sources[ $current_mime ] = array(
+			'file'     => array_key_exists( 'file', $current_size ) ? $current_size['file'] : '',
+			'filesize' => 0,
+		);
+
+		// Set the filesize from the current mime image.
+		$file_location = path_join( $dirname, $sources[ $current_mime ]['file'] );
+		if ( file_exists( $file_location ) ) {
+			$sources[ $current_mime ]['filesize'] = filesize( $file_location );
+		}
+
+		$formats = array_key_exists( $current_mime, $valid_mime_transforms )
+			? $valid_mime_transforms[ $current_mime ]
+			: array();
+
+		foreach ( $formats as $mime ) {
+			wp_schedule_single_event( time(), 'webp_uploads_create_image', array( $attachment_id, $size_name, $mime ) );
+		}
```

Comment:
The code here is still missing the critical check for whether we actually need to generate the format still. Since the process may fail at any point, we need to always make sure we only generate the sizes which are missing, see also my feedback in https://github.com/WordPress/performance/pull/147#issuecomment-1043145067.

Also, why a cron event? WordPress core doesn't use a cron event to generate its image sizes, and I think we should follow how it generates its default sub-sizes, which happens in a single request, just allowing for failures as far as I can tell, and then it retries until everything has been completed (see the `_wp_make_subsizes()` function and how it is used).

We also need to make sure then that we keep using the latest version of `$sources`, since it could be updated within the function, so your `webp_uploads_generate_image_size()` function below should probably return the source array that it has added.

How about this:
```suggestion
		foreach ( $formats as $mime ) {
			if ( isset( $sources[ $mime ] ) ) {
				continue;
			}
			$new_source = webp_uploads_generate_image_size( $attachment_id, $size_name, $mime );
			if ( is_wp_error( $new_source ) ) {
				continue;
			}
			$sources[ $mime ] = $new_source;
		}
```

Reply:
Actually, this is already covered here: 

https://github.com/WordPress/performance/pull/147/files/6e3afeef0648073c2615aa15f4d0e090f0b3b93f#diff-71ed188a16b79d179868ee80e76f775462a451e871fc3d6214fcacca3edcea8cR47

As if the size has not been created yet the additional mime types for that image won't be created. 

> Since the process may fail at any point, we need to always make sure we only generate the sizes which are missing, see also my feedback in https://github.com/WordPress/performance/pull/147#issuecomment-1043145067.

Once the image is retried and finally generated it would trigger this hook again, which at that point would contain the key for that particular image and then the image would be scheduled for creation. 

> Also, why a cron event? 

To split the work for every single size in a different process, and only use as minimal resources as possible per image size so we can decrease the number of potential failures when creating an image in a constrained environment.

> WordPress core doesn't use a cron event to generate its image sizes

It does not because it has to process the image in a single request and display the image right away, we as progressive enchantment can do that as we know a fallback image was already created. And we can create a background process (Cron) for any additional mime images.

This mechanism would ensure existing users would continue to use the same number of resources when uploading an image as they currently do, without the overhead of additional resources introduced by a new mime type. 

> just allowing for failures as far as I can tell, and then it retries until everything has been completed (see the _wp_make_subsizes() function and how it is used).

The failures would be handled by this hook as described at the beginning every time an image fails won't be added into the image sizes, and every time a new image size is added it would trigger this hook, ensuring that the image is only scheduled for creation when the main image was created.

> We also need to make sure then that we keep using the latest version of $sources, since it could be updated within the function, so your webp_uploads_generate_image_size() function below should probably return the source array that it has added.

This is already handled here by just returning the updated metadata and the hook dispatched by this function would update the meta once is executed. 

- https://github.com/WordPress/performance/pull/147/files/6e3afeef0648073c2615aa15f4d0e090f0b3b93f#diff-71ed188a16b79d179868ee80e76f775462a451e871fc3d6214fcacca3edcea8cR103

Reply:
Although the `wp_schedule_single_event` approach works in my testing, I wonder if it will work across WordPress installs or might create problems for some users, for example does this work if `define( 'DISABLE_WP_CRON', true ); ` is set?

I think we should test Felix's suggestion of trying to generate all images in one go, tracking the completed ones in meta . Core's upload process (in both media and the block editor) will catch incomplete generation runs (a timeout error) and send a retry request (which we may need to hook into?). I will try testing this approach with a low server timeout to verify.

Reply:
i encountered a big with the callback approach: after uploading an image, I deleted it before WebP creation completed. A final WebP was generated after the deletion.

testing with a short server timeout, WordPress's retry mechanism worked as expected, and the cron callback was set up after the core image process completed.

Reply:
> Although the wp_schedule_single_event approach works in my testing, I wonder if it will work across WordPress installs or might create problems for some users, for example does this work if define( 'DISABLE_WP_CRON', true ); is set?

Usually when this is the case, is to run the corn with an actual CRON via the server which will still execute any remaining cron job. However, if the CRON is completely disabled from the server any additional functionality from WordPress won't work either like: 

- Scheduling publishing 
- Automatic updates from WP and plugins.

> I think we should test Felix's suggestion of trying to generate all images in one go, tracking the completed ones in meta . Core's upload process (in both media and the block editor) will catch incomplete generation runs (a timeout error) and send a retry request (which we may need to hook into?). I will try testing this approach with a low server timeout to verify.

Can you share your setup PHP.ini so I can try to replicate it?

> i encountered a big with the callback approach: after uploading an image, I deleted it before WebP creation was completed. A final WebP was generated after the deletion.

Oh, great point I think this can be an edge case but still would be ideal to make sure any remaining cron is removed when the image is removed as well.

Reply:
@mitogh 
> > Also, why a cron event?
>
> To split the work for every single size in a different process, and only use as minimal resources as possible per image size so we can decrease the number of potential failures when creating an image in a constrained environment.

While I get the thinking behind this, we have to keep in mind we're building this for WordPress core, which already has an existing working solution for generating sub-sizes in a scalable way. WP core does not use cron for this, and if we wanted to change that, it would open up a big discussion on whether that makes sense or not, which is not the point we're trying to make here - we want to add support for adding sub-sizes also in WebP.

We should really minimize any sort of "refactoring" that is unrelated to the problem we're trying to solve. We can tie into the existing logic and follow the same approach that WP core has here without running into problems - and as @adamsilverstein has pointed out, using a different approach here potentially causes other problems and it will require a lot of extra testing - in contrast to using WP core's existing approach, which we know works.

For the above reasons, I still think it would be better to drop the cron usage here and just call the function right away, as that follows the existing approach in WP core.

Reply:
I did some testing with a low php timeout threshold using both approaches. Unfortunately, the way we hook in now is _after core completes its retries_, so our generation code only gets one run. With a low timeout, this means not all sizes will be generated. We could either change how we hook in so it happens earlier (before completion so the error is caught), or we could hook in later and return an error if mime/sizes are still missing:

For Ajax hooking in with the `wp_prepare_attachment_for_js` [filter](url), which is the last thing the ajax callback [calls](https://github.com/WordPress/wordpress-develop/blob/74d7ef73b361fff82605291282588dc05059dc27/src/wp-admin/includes/ajax-actions.php#L2476) (returning false when mime/sizes are missing would trigger another retry)

I found the same issue in Gutenberg, the retry 'create-image-subsizes' was only called until all the jpg images are generated, then one attempt at webp images, the retry did not continue because the [code](https://github.com/WordPress/gutenberg/blob/ea471c0161a3e3bb9af7424c85985f7b8be8aa67/packages/api-fetch/src/middlewares/media-upload.js#L45-L60) only retries until a success return. We may be able to use the `rest_prepare_attachment` filter (in `src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php::L843). Returning an Error when the attachment has missing mime/sizes woulds trigger the retry.

Maybe we can also bump these retries up on the trac ticket.

Reply:
I'm fine merging Felix's suggested code, then working on ensuring error catching (and retries) work correctly for secondary types as well. 

It might be worth looking at hooking on `wp_update_attachment_metadata`, which is called for each image core generates, passing the full meta with each call, with the last item in the sizes array the most recently generated image. If our image generation threw a timeout error here it could trigger the retry.

`wp_update_attachment_metadata` is also called by image edits which could be handy if we want to handle those in alternate mime types.

Reply:
@adamsilverstein Based on your feedback, I agree we may need to hook into `wp_update_attachment_metadata`. That would require some tweaks for our callback though, since that hook is fired every time a single new sub-size is generated in core. We had also discussed the potential usage of this hook in our call last week - the tricky part would be to figure out whether it is used for an image size being generated so that our code is only run in that case. Also, the logic for which sub-sizes to generate may need to be reworked.

Before we go towards doing that though, are we sure the current implementation is not working in a scalable way? I understand your argumentation @adamsilverstein, but why is that actually failing? Looking at the code, WP core uses `wp_update_image_subsizes()`, both via AJAX and the REST API. Before that function succeeds, it fires the `wp_generate_attachment_metadata` filter, which we are using here, so I don't follow how the code would return a success response without running our added filter callback 🤔 

In any case, I think it makes sense to drop the cron hook, replace it with a direct call, and explore further from there on whether the `wp_generate_attachment_metadata` filter is okay or whether we need an alternative solution using the `wp_generate_attachment_metadata` filter. cc @mitogh

Reply:
Summarizing a chat with @mitogh and @adamsilverstein here:
* We're going to switch to "direct" generation of sub-sizes, essentially replace the cron hook with a direct call to the function as suggested above.
* The remaining problem with that approach is that `wp_update_image_subsizes()` will no longer trigger the `wp_generate_attachment_metadata` filter in any subsequent requests after the WP core sizes have been generated (since `wp_get_missing_image_subsizes()` returns an empty array, it short-circuits) - that's why the implementation will only make one more request after a failure in _our_ code, but that won't trigger the hook, so WP thinks the sub-sizes generation has succeeded.
* We can work around that by hooking into the `wp_get_missing_image_subsizes` filter, and essentially misusing it as an action, as follows:
    * If the list of sizes is empty _and_ we are in some sort of "upload" request, we manually call our function to generate the missing sub-sizes in the additional formats.
    * We just return the original filtered value without changing it.
    * In order to know whether we are in an "upload" request, we have to set a temporary meta key in the attachment as soon as our logic triggers, to flag that this attachment is basically "awaiting additional image MIME sub-sizes generation". Once our function has completed, we remove that same meta key flag again.
    * This approach will work regardless of whether image sizes are being generated via WP-AJAX or REST.
    * It is quite hacky, but it's what we need to do to work with the limits of not having a proper filter/action in WP core to use here. We should add a comment to that code indicating how that would behave differently in an eventual WP core patch.
* Let's limit this PR to going with just the simple removal of the cron hook. The above "hack" we can then add in a follow-up PR to the same issue #142 - that allows the review to be simpler, since already a lot of work has gone into this, and it's a good stage to break the work into two PRs.